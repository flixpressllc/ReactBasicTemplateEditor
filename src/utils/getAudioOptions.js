import {nativeXmlToObject} from './xmlAdapter';

// https://ws.flixpress.com/AudioWebService.asmx/GetAudio?categoryId=2&page=1&pageSize=100&username=bowdo

// https://ws.flixpress.com/AudioWebService.asmx/GetCategoryTree

const CATEGORY_URL = 'https://ws.flixpress.com/AudioWebService.asmx/GetCategoryTree';
const CUSTOM_URL = 'https://ws.flixpress.com/CustomAudioWebService.asmx/GetCustomAudio';
const AUDIO_URL = 'https://ws.flixpress.com/AudioWebService.asmx/GetAudio';

function whenAll(arrayOfPromises) {
  return $.when.apply($, arrayOfPromises).then(function() {
    return Array.prototype.slice.call(arguments, 0);
  });
}

function getCatSongs(categoryId, username) {
  return new Promise((res, rej) => {
    $.ajax({
      url: AUDIO_URL,
      data: {username: username, categoryId: categoryId, page:1, pageSize: 1000},
      dataType: 'xml',
      type: 'GET'
    }).done(function (result) {
      let songs = nativeXmlToObject(result).ResultSetOfAudio.Records.Audio;
      res(songs);
    }).fail(err => rej(err));
  });
}

function getCustomSongsForUser (username) {
  return new Promise((res, rej) => {
    $.ajax({
      url: CUSTOM_URL,
      dataType: 'xml',
      type: 'GET',
      data: {username: username, page:1, pageSize: 1000}
    }).done(function (result) {
      var songs = nativeXmlToObject(result).ResultSetOfCustomAudio.Records.CustomAudio;
      songs = songs === undefined ? [] : songs;
      res(songs);
    }).fail(function (err) { rej(err) });
  });
}

function getCategoriesAndSongsForUser (username) {
  return new Promise((res, rej) => {
    $.ajax({
      url: CATEGORY_URL,
      dataType: 'xml',
      type: 'GET'
    }).done(function(result){
      var getAllCats = [];
      var categoriesObj = {};
      var categories = nativeXmlToObject(result).ArrayOfCategory.Category.SubCategories.Category;

      $.each(categories, function(arrPos, category){
        var catSongs = getCatSongs(category.Id, username)
        getAllCats.push(catSongs);
        catSongs.then(function(songs){
          categoriesObj[category.Name] = {};
          categoriesObj[category.Name].id = category.Id;
          categoriesObj[category.Name].songs = songs;
        });
      });

      Promise.all(getAllCats).then(function(){
        res(categoriesObj);
      }).catch(function (err) {
        rej(err);
      })
    }).fail(function (err) {
      rej(err);
    });
  });
}

export default function getAudioOptions (username) {
  var everythingIsReady = $.Deferred();

  var haveAllCats = $.Deferred();
  var haveAllCustom = $.Deferred();

  getCategoriesAndSongsForUser(username).done(catsObj => {
    haveAllCats.resolve(catsObj)
  })

  getCustomSongsForUser(username).then(songs => {
    haveAllCustom.resolve(songs);
  })

  $.when(haveAllCats, haveAllCustom).then(function (categories, customAudio) {
    everythingIsReady.resolve({categories, customAudio});
  });

  return everythingIsReady;
}
