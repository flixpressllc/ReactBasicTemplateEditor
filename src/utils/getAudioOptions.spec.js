import getAudioOptions from './getAudioOptions';

global.window.$ = require('jquery');

describe('getAutioOptions', () => {
  describe('custom audio', () => {
    it('returns an empty array if custom audio is unavailable', () => {
      pending();
      getAudioOptions.getAudioOptions('DonDenton');
    });
  });

  it('returns a promise that responds to done', () => {
    const done = getAudioOptions('bowdo').done
    expect(done).not.toEqual(undefined);
  });

  it('yields the expected data in done', () => {
    const p = new Promise(res => {
      getAudioOptions('DonDenton').done(data => res(data))
    });
    return p.then(results => {
      expect(results).toMatchSnapshot();
    })
  });
});
