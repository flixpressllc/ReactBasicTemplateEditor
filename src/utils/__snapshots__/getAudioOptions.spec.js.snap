exports[`getAutioOptions yields the expected data in done 1`] = `
Object {
  "categories": Object {
    "Classical": Object {
      "id": 2,
      "songs": Array [
        Object {
          "Id": 2,
          "Keywords": "Anitra",
          "Length": 2,
          "LicensedAudioAccessTypeId": 1,
          "Name": "Anitra",
        },
        Object {
          "Id": 3,
          "Keywords": "Bumblebee",
          "Length": 2,
          "LicensedAudioAccessTypeId": 2,
          "Name": "Bumblebee",
        },
        Object {
          "Id": 4,
          "Keywords": "Danube",
          "Length": 2,
          "LicensedAudioAccessTypeId": 2,
          "Name": "Danube",
        },
        Object {
          "Id": 7,
          "Keywords": "Morning",
          "Length": 2,
          "LicensedAudioAccessTypeId": 2,
          "Name": "Morning",
        },
        Object {
          "Id": 8,
          "Keywords": "Moutain King",
          "Length": 2,
          "LicensedAudioAccessTypeId": 2,
          "Name": "Moutain King",
        },
      ],
    },
    "Cute-Lite": Object {
      "id": 3,
      "songs": Array [
        Object {
          "Id": 14,
          "Keywords": "Everyday Life",
          "Length": 3,
          "LicensedAudioAccessTypeId": 1,
          "Name": "Everyday Life",
        },
        Object {
          "Id": 15,
          "Keywords": "Fairy Toes",
          "Length": 3,
          "LicensedAudioAccessTypeId": 2,
          "Name": "Fairy Toes",
        },
        Object {
          "Id": 16,
          "Keywords": "Happy Product",
          "Length": 3,
          "LicensedAudioAccessTypeId": 2,
          "Name": "Happy Product",
        },
        Object {
          "Id": 18,
          "Keywords": "Hi Tech Evo",
          "Length": 3,
          "LicensedAudioAccessTypeId": 2,
          "Name": "Hi Tech Evo",
        },
        Object {
          "Id": 23,
          "Keywords": "Sunshine",
          "Length": 3,
          "LicensedAudioAccessTypeId": 2,
          "Name": "Sunshine",
        },
      ],
    },
    "High-Energy": Object {
      "id": 4,
      "songs": Array [
        Object {
          "Id": 1,
          "Keywords": "5th Symphony",
          "Length": 4,
          "LicensedAudioAccessTypeId": 1,
          "Name": "5th Symphony",
        },
        Object {
          "Id": 26,
          "Keywords": "Breakdown",
          "Length": 4,
          "LicensedAudioAccessTypeId": 1,
          "Name": "Breakdown",
        },
        Object {
          "Id": 27,
          "Keywords": "Breakthrough",
          "Length": 4,
          "LicensedAudioAccessTypeId": 2,
          "Name": "Breakthrough",
        },
        Object {
          "Id": 31,
          "Keywords": "Forward Drive",
          "Length": 4,
          "LicensedAudioAccessTypeId": 2,
          "Name": "Forward Drive",
        },
        Object {
          "Id": 33,
          "Keywords": "My Choice",
          "Length": 4,
          "LicensedAudioAccessTypeId": 2,
          "Name": "My Choice",
        },
        Object {
          "Id": 37,
          "Keywords": "Tech Revolution",
          "Length": 4,
          "LicensedAudioAccessTypeId": 2,
          "Name": "Tech Revolution",
        },
      ],
    },
    "Inspirational": Object {
      "id": 5,
      "songs": Array [
        Object {
          "Id": 40,
          "Keywords": "Driving Mentality",
          "Length": 5,
          "LicensedAudioAccessTypeId": 1,
          "Name": "Driving Mentality",
        },
        Object {
          "Id": 44,
          "Keywords": "Inspiring Piano",
          "Length": 5,
          "LicensedAudioAccessTypeId": 2,
          "Name": "Inspiring Piano",
        },
        Object {
          "Id": 46,
          "Keywords": "Morning Shine",
          "Length": 5,
          "LicensedAudioAccessTypeId": 2,
          "Name": "Morning Shine",
        },
        Object {
          "Id": 49,
          "Keywords": "Winning Circle",
          "Length": 5,
          "LicensedAudioAccessTypeId": 2,
          "Name": "Winning Circle",
        },
      ],
    },
    "Promo": Object {
      "id": 6,
      "songs": Array [
        Object {
          "Id": 50,
          "Keywords": "Above All",
          "Length": 6,
          "LicensedAudioAccessTypeId": 1,
          "Name": "Above All",
        },
        Object {
          "Id": 52,
          "Keywords": "Ashes 2 Ashes",
          "Length": 6,
          "LicensedAudioAccessTypeId": 2,
          "Name": "Ashes 2 Ashes",
        },
        Object {
          "Id": 54,
          "Keywords": "Dance With Me",
          "Length": 6,
          "LicensedAudioAccessTypeId": 2,
          "Name": "Dance With Me",
        },
        Object {
          "Id": 63,
          "Keywords": "True Lies",
          "Length": 6,
          "LicensedAudioAccessTypeId": 2,
          "Name": "True Lies",
        },
        Object {
          "Id": 64,
          "Keywords": "Western Sun",
          "Length": 6,
          "LicensedAudioAccessTypeId": 2,
          "Name": "Western Sun",
        },
      ],
    },
    "Short-Intro": Object {
      "id": 7,
      "songs": Array [
        Object {
          "Id": 65,
          "Keywords": "Intro 1",
          "Length": 7,
          "LicensedAudioAccessTypeId": 1,
          "Name": "Intro 1",
        },
        Object {
          "Id": 66,
          "Keywords": "Intro 2",
          "Length": 7,
          "LicensedAudioAccessTypeId": 1,
          "Name": "Intro 2",
        },
        Object {
          "Id": 67,
          "Keywords": "Intro 3",
          "Length": 7,
          "LicensedAudioAccessTypeId": 1,
          "Name": "Intro 3",
        },
        Object {
          "Id": 68,
          "Keywords": "Intro 4",
          "Length": 7,
          "LicensedAudioAccessTypeId": 1,
          "Name": "Intro 4",
        },
        Object {
          "Id": 69,
          "Keywords": "Intro 5",
          "Length": 7,
          "LicensedAudioAccessTypeId": 1,
          "Name": "Intro 5",
        },
        Object {
          "Id": 70,
          "Keywords": "Intro 6",
          "Length": 7,
          "LicensedAudioAccessTypeId": 1,
          "Name": "Intro 6",
        },
        Object {
          "Id": 71,
          "Keywords": "Intro 7",
          "Length": 7,
          "LicensedAudioAccessTypeId": 1,
          "Name": "Intro 7",
        },
        Object {
          "Id": 72,
          "Keywords": "Intro 8",
          "Length": 7,
          "LicensedAudioAccessTypeId": 1,
          "Name": "Intro 8",
        },
      ],
    },
    "Various": Object {
      "id": 8,
      "songs": Array [
        Object {
          "Id": 73,
          "Keywords": "Crossing Over",
          "Length": 8,
          "LicensedAudioAccessTypeId": 1,
          "Name": "Crossing Over",
        },
        Object {
          "Id": 74,
          "Keywords": "Dancey",
          "Length": 8,
          "LicensedAudioAccessTypeId": 2,
          "Name": "Dancey",
        },
      ],
    },
  },
  "customAudio": Array [
    Object {
      "Filename": 1152,
      "Id": 1152,
      "Length": 186331,
      "Name": "My Audio",
      "UserId": 1075982,
    },
    Object {
      "Filename": 3442,
      "Id": 3442,
      "Length": 2351,
      "Name": "test voice",
      "UserId": 1075982,
    },
  ],
}
`;
